odl/tomo/data/uncompr_bin.py

```diff
@@ -0,0 +1,472 @@
+            raise ValueError('`header_bytes` is larger or equal to file size '
+                             '({} >= {})'.format(header_bytes, filesize_bytes))
+
+        if dtype is None:
+            raise TypeError('`dtype` cannot be `None`')
+        dtype = np.dtype(dtype)
+        data_size = data_bytes / dtype.itemsize
+
+        instance = cls(f, header_fields=[])
+        instance.data_dtype = dtype
+        instance.data_shape = (data_size,)
+        instance.header_bytes = header_bytes
+        return instance
+
+    def read_header(self):
+        """Read the header from the reader's file.
+
+        The header is also stored in the ``self.header`` attribute.
+
+        Returns
+        -------
+        header : dict
+            Header of ``self.file`` stored in a dictionary, where each
+            entry has the following form::
+
+                'name': {'value': value, 'description': description}
+
+            All ``value``'s are `numpy.ndarray`'s with at least one
+            dimension. If a ``shape`` is given in ``self.header_fields``,
+            the resulting array is reshaped accordingly.
+        """
+        # Read all fields except data
+        header = {}
+        for field in self.header_fields:
+            # Get all the values from the dictionary
+            name = field['name']
+            if name == 'data':
+                continue
+            entry = {'description': field.get('description', '')}
+            offset_bytes = field['offset_bytes']
+            size_bytes = field.get('size_bytes', None)
+            dtype = field['dtype']
+            shape = field.get('dshape', None)
+            if shape is None:
+                shape = -1  # Causes reshape to 1d or no-op
+
+            bytes_per_elem = dtype.itemsize
+
+            if size_bytes is None:
+                # Default if 'size_bytes' is omitted
+                num_elems = 1
+            else:
+                num_elems = size_bytes / bytes_per_elem
+
+            if num_elems != int(num_elems):
+                raise RuntimeError(
+                    "field '{}': `size_bytes` {} and `dtype.itemsize` {} "
+                    " result in non-integer number of elements"
+                    "".format(name, size_bytes, bytes_per_elem))
+
+            # Create format string for struct module to unpack the binary
+            # data
+            fmt = str(int(num_elems)) + DTYPE_MAP_NPY2STRUCT[dtype]
+            if struct.calcsize(fmt) != size_bytes:
+                raise RuntimeError(
+                    "field '{}': format '{}' has results in {} bytes, but "
+                    "`size_bytes` is {}"
+                    "".format(name, fmt, struct.calcsize(fmt), size_bytes))
+
+            self.file.seek(offset_bytes)
+            packed_value = self.file.read(size_bytes)
+            value = np.array(struct.unpack_from(fmt, packed_value),
+                             dtype=dtype)
+
+            if dtype == np.dtype('S1'):
```

Comment:
For me this is the simplest way of making the string case consistent with the other cases. The specification says "4 bytes of type Char", which is equivalent to `np.array(..., dtype='S1').shape == (4,)`. I may need to not `struct.unpack` them if they are to be read as strings. Some testing required.